packages/components/modal/src/Modal.tsx

```diff
@@ -143,19 +143,32 @@ export const Modal = ({
     overlayClassName: otherProps.overlayProps?.className,
   });
 
-  React.useEffect(() => {
-    if (props.isShown) {
-      setTimeout(() => {
-        if (props.initialFocusRef && props.initialFocusRef.current) {
-          if (props.initialFocusRef.current.focus) {
-            props.initialFocusRef.current.focus();
-          }
-        } else if (contentRef.current) {
-          focusFirstWithinNode(contentRef.current);
-        }
-      }, 100);
+  const onInitialFocus = () => {
+    const contentEl = contentRef.current;
+    if (contentEl) {
+      const activeEl = document.activeElement;
+      const isContentContainsActive =
+        contentEl !== activeEl && contentEl.contains(activeEl);
+
+      if (isContentContainsActive) {
+        return;
+      }
+    }
+
+    const initialFocusEl = props.initialFocusRef?.current;
+    if (initialFocusEl) {
+      initialFocusEl.focus?.();
+    } else if (contentEl) {
+      focusFirstWithinNode(contentEl);
     }
```

Comment:
Knowing the following (snippets from MDN), we can simplify a bit:

`activeElement`:
> The [Element](https://developer.mozilla.org/en-US/docs/Web/API/Element) which currently has focus, [<body>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/body) or null if there is no focused element.

`Node.contains`:
> If the otherNode parameter is null, contains() always returns false.

The following should work since `focusFirstWithinNode` is safe to call with a null argument. I removed the check for `contentEl !== activeEl`, though perhaps I'm missing an edge case where that's possible?

```suggestion
    const contentEl = contentRef.current;
    if (contentEl?.contains(document.activeElement)) {
      // A descendant of the modal already has focus.
      return;
    }

    const initialFocusEl = props.initialFocusRef?.current;
    if (initialFocusEl) {
      initialFocusEl.focus?.();
      return;
    }
    
    focusFirstWithinNode(contentEl);
```

Reply:
> I removed the check for contentEl !== activeEl, though perhaps I'm missing an edge case where that's possible?

the edge case here is that `contentEl` becomes initially focused by `react-modal`